ILogger_best_practice/kb/scenario_description/defaults.md

```diff
@@ -0,0 +1,182 @@
+# Defaults
+
+## String interpolation using the $ symbol
+
+String interpolation in C# is a feature that allows for the inclusion of variable values directly within a string literal. This is achieved using the $ symbol followed by a string enclosed in double quotes. Inside this string, expressions can be embedded within curly braces {}, and these expressions are evaluated and their values are included in the resulting string.
+
+### Syntax
+
+The basic syntax of string interpolation using the $ symbol is as follows:
+
+```csharp
+string interpolatedString = $"This is a string with an embedded value: {expression}";
+```
+
+### Examples
+
+#### Basic Example
+
+```csharp
+int age = 25;
+string name = "Alice";
+string message = $"My name is {name} and I am {age} years old.";
+Console.WriteLine(message);
+// Output: My name is Alice and I am 25 years old.
+```
+
+#### Expressions within Interpolations
+
+```csharp
+int a = 5;
+int b = 10;
+string result = $"The sum of {a} and {b} is {a + b}.";
+Console.WriteLine(result);
+// Output: The sum of 5 and 10 is 15.
+```
+
+#### Calling Methods and Properties
+
+```csharp
+DateTime now = DateTime.Now;
+string formattedDate = $"Today's date is {now:MMMM dd, yyyy}.";
+Console.WriteLine(formattedDate);
+// Output: Today's date is (current date in "MMMM dd, yyyy" format).
+```
+
+#### Using with Object Properties
+
+```csharp
+var person = new { Name = "John", Age = 30 };
+string introduction = $"Hello, my name is {person.Name} and I am {person.Age} years old.";
+Console.WriteLine(introduction);
+// Output: Hello, my name is John and I am 30 years old.
+```
+
+## Compile-time logging source generation
+
+.NET 6 introduces the LoggerMessageAttribute type. This attribute is part of the `Microsoft.Extensions.Logging` namespace, and when used, it source-generates performant logging APIs.
+The source-generation logging support is designed to deliver a highly usable and highly performant logging solution for modern .NET applications. The auto-generated source code relies on the `Microsoft.Extensions.Logging.ILogger` interface.
+
+The source generator is triggered when LoggerMessageAttribute is used on `partial` logging methods. When triggered, it is either able to autogenerate the implementation of the partial methods it's decorating, or produce compile-time diagnostics with hints about proper usage. The compile-time logging solution is typically considerably faster at run time than existing logging approaches. It achieves this by eliminating boxing, temporary allocations, and copies to the maximum extent possible.
```

Comment:
I don't know how much it matters to readers but I believe the biggest performance improvement for source-generated logging is the one that often gets overlooked. When code calls LogInformation("Could not open socket to `{HostName}`", hostname), a bunch of work happens to parse and transform the string "Could not open socket to `{HostName}`" into a formatter delegate `Func<TState,Exception?,string>`. The result gets stored in a cache to hopefully speed up future log calls but that cache has a limited size. Once that cache is full every Log call that uses a template string not in the cache will have to pay the full cost of that parse and transform operation.

Source-generated logging avoids that problem because all the parsing and formatting logic gets generated at build time.